stock_picking_invoicing/wizards/stock_invoice_onshipping.py

```diff
@@ -0,0 +1,553 @@
+            key = (
+                picking._get_partner_to_invoice(),
+                picking.picking_type_id,
+                picking.partner_id,
+            )
+        return key
+
+    def _group_pickings(self, pickings):
+        """
+        Group given picking
+        :param pickings:
+        :return: list of stock.picking recordset
+        """
+        grouped_picking = {}
+        pickings = pickings.filtered(lambda p: p.invoice_state == "2binvoiced")
+        for picking in pickings:
+            key = self._get_picking_key(picking)
+            picks_grouped = grouped_picking.get(key, self.env["stock.picking"].browse())
+            picks_grouped |= picking
+            grouped_picking.update({key: picks_grouped})
+        return grouped_picking.values()
+
+    def _simulate_invoice_onchange(self, values):
+        """
+        Simulate onchange for invoice
+        :param values: dict
+        :return: dict
+        """
+        invoice = self.env["account.move"].new(values.copy())
+        invoice._onchange_partner_id()
+        new_values = invoice._convert_to_write(invoice._cache)
+        # Ensure basic values are not updated
+        values.update(new_values)
+        return invoice, values
+
+    def _build_invoice_values_from_pickings(self, pickings):
+        """
+        Build dict to create a new invoice from given pickings
+        :param pickings: stock.picking recordset
+        :return: dict
+        """
+        picking = fields.first(pickings)
+        partner_id = picking._get_partner_to_invoice()
+        partner = self.env["res.partner"].browse(partner_id)
+        inv_type = self._get_invoice_type()
+        if inv_type in ("out_invoice", "out_refund"):
+            payment_term = partner.property_payment_term_id.id
+        else:
+            payment_term = partner.property_supplier_payment_term_id.id
+        company = self.env.company
+        currency = company.currency_id
+        if partner:
+            code = picking.picking_type_id.code
+            if partner.property_product_pricelist and code == "outgoing":
+                currency = partner.property_product_pricelist.currency_id
+        journal = self._get_journal()
+        invoice_obj = self.env["account.move"]
+        values = invoice_obj.default_get(invoice_obj.fields_get().keys())
+        values.update(
+            {
+                "invoice_origin": ", ".join(pickings.mapped("name")),
+                "user_id": self.env.user.id,
+                "partner_id": partner_id,
+                "invoice_payment_term_id": payment_term,
+                "move_type": inv_type,
+                "fiscal_position_id": partner.property_account_position_id.id,
+                "company_id": company.id,
+                "currency_id": currency.id,
+                "journal_id": journal.id,
+                "picking_ids": [(4, p.id, False) for p in pickings],
+            }
+        )
```

Comment:
What is the idea behind defining user_id, invoice_payment_term_id, fiscal_position_id and company_id because those fields are already computed by odoo with the onchange ?